packages/theme/src/cli/utilities/theme-fs.ts

```diff
@@ -210,16 +204,22 @@ export function mountThemeFileSystem(root: string, options?: ThemeFileSystemOpti
         return false
       })
 
-    emitEvent('unlink', {
-      fileKey,
-      onSync: (fn) => {
-        syncPromise
-          .then((didSync) => {
-            if (didSync) fn()
-          })
-          .catch(() => {})
-      },
-    })
+    // We pass the syncPromise to the 'unlink' event for liquid assets so we can defer triggering the hot reload
+    // until the file has been deleted on the server
+    if (fileKey.startsWith('assets/') && extname(fileKey) === '.liquid') {
+      emitEvent('unlink', {
+        fileKey,
+        onSync: (fn) => {
+          syncPromise
+            .then((didSync) => {
+              if (didSync) fn()
+            })
+            .catch(() => {})
+        },
+      })
+    } else {
+      emitEvent('unlink', {fileKey})
+    }
```

Comment:
To my understanding, this is just making the `onSync` property available only for liquid assets, right?
Does it hurt to make it available to every asset, though?

I think the logic that decides when to do the hot reload is already in place in `hot-reload/server`, perhaps we shouldn't leak it here?